### PR TITLE
[Hexagon] Do not auto-build apps when building TVM

### PR DESCRIPTION
**Pull request by kparzysz-quic**
The Hexagon cmakes have recently become unwieldy due to a complex network of dependencies between various automatically built components. This was in large part because of trying to automatically build some apps, which then tried to build TVM runtimes again, but with their own configurations.

This patch removes the ability to automatically build any Hexagon-related apps from the main TVM build. The following cmake options are now deprecated:
  - `USE_HEXAGON_LAUNCHER`
  - `USE_HEXAGON_PROXY_RPC`

In order to build the binaries needed for HexagonLauncher from `tvm.contrib.hexagon`:
  - Build TVM+runtime for x86, with codegen for Hexagon enabled. This can be done via `USE_HEXAGON_DEVICE=sim` or `target`.
  - Build Android runtime and tvm_rpc with `-DUSE_RPC=ON`, `-DUSE_CPP_RPC=ON`, and `-DUSE_HEXAGON_RPC=ON`.
  - Build Hexagon runtime with `-DUSE_HEXAGON_RPC=ON`, and `-DBUILD_STATIC_RUNTIME=ON`.

Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

### CODE REVIEW

**Comment by kparzysz-quic** [comment]
cc: @mehrdadh

**Comment by Lunderberg** [comment]
I like the idea of simplifying the build process, but I'm worried that this would make it more difficult for developers to test changes.  This PR would reduce the complexity of the existing cmake setup, but I think it would result in passing that complexity to the developer to manually handle.

Build details, as I understand them.

* Three TVM builds are required.  (Host, connected Android, connected Hexagon)
* Each of the three builds has different values for the RPC and static/dynamic options.
* All three builds must have the same values for several build variables.
  * Hexagon: `USE_HEXAGON_SDK` and `USE_HEXAGON_ARCH`
  * Android: `USE_ANDROID_TOOLCHAIN`
* All three builds must have be built from the same version of TVM.
* All three builds must output to the same location (currently `$BUILD_DIR/apps/hexagon_*`).

Brainstorming on options

* Option 1: Top-level TVM CMake launches the three independent builds as external projects
  * Current implementation on main.
  * Pro: Developer can run `make && run_my_test.py` to rebuild all three tests in a single step.
  * Con: Difficult to manage setup.
* Option 2: Developer launches the three independent builds
  * Current implementation in this PR.
  * Pro: Easier to manage cmake setup.
  * Con: Interactions between the three builds must be handled by each developer, either manually or by writing a wrapper build script.
  * Con: Removes the app from the TVM framework.  Compiling TVM is no longer sufficient to use `tvm.contrib.hexagon`.
* Option 3: App Cmake launches all three independent builds
  * Pro: Developer can run tests from a single location
  * Pro: Includes occur only in one direction, from the app to the TVM build.
  * Con: Removes the app from the TVM framework.  Compiling TVM is no longer sufficient to use `tvm.contrib.hexagon`.

Between the high-level options, I'd prefer Option 1, since it would maintain the same developer-facing interactions.  Thoughts?

**Comment by kparzysz-quic** [comment]
I chatted with Mehrdad and we're going with an app that will build the Hexagon RPC support.

File: apps/hexagon_api/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+# Required variables:
+#   ANDROID_ABI
+#   ANDROID_PLATFORM
+#   USE_ANDROID_TOOLCHAIN (Android toolchain .cmake file)
+#   USE_HEXAGON_ARCH
+#   USE_HEXAGON_SDK
+#   USE_HEXAGON_TOOLCHAIN (Path to Hexagon toolchain ending with "Tools")
+
+set(TVM_SOURCE_DIR "${CMAKE_SOURCE_DIR}/../..")
+set(HEXAGON_API_BINARY_DIR "${CMAKE_BINARY_DIR}/hexagon_rpc")
```
Comment (mehrdadh) [inline apps/hexagon_api/CMakeLists.txt:16]: change HEXAGON_API_BINARY_DIR to this to be compatible with current HexagonLauncher design:
set(HEXAGON_API_BINARY_DIR "${CMAKE_BINARY_DIR}/../../../build/hexagon_rpc")

This is with the assumption of building this CmakeLists.txt under `apps/hexagon_api/build`. Not the best solution, but we need to make sure everything is copied to `build/hexagon_rpc`.

File: apps/hexagon_api/README.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+The build will create a subdirectory `hexagon_rpc` with the following
+binaries
+- `tvm_runtime.so`: TVM runtime for Android (shared library).
+- `tvm_rpc_android`: RPC server for Android.
+- `libhexagon_rpc_skel.so`: RPC library for Hexagon.
+- `libtvm_runtime.a`: TVM runtime for Hexagon (static library).
+
+The RPC library for Hexagon contains the TVM runtime, so the static
+TVM runtime for Hexagon is not strictly necessary.
```
Comment (mehrdadh) [inline apps/hexagon_api/README.md:54]: I suggest to add an instruction like this to make sure we keeps the relative directory to output of binary files.

```
## Instruction

cd apps/hexagon_api
mkdir build
cd build
cmake -DConfig1 -DConfig2 ..
```

**Review by mehrdadh — CHANGES_REQUESTED** [review]
Thanks for making this change! Just few minor things to make sure tests are passing.

File: apps/hexagon_api/CMakeLists.txt (hunk 1 above)
Comment (kparzysz-quic) [reply apps/hexagon_api/CMakeLists.txt:16]: This is going to be really fragile.  Instead I propose a flag `USE_OUTPUT_BINARY_DIR` that could be set to any location, including `.../build/hexagon_rpc`.

File: apps/hexagon_api/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
```
[the remark this answers is not in the document]
Comment (kparzysz-quic) [reply apps/hexagon_api/CMakeLists.txt:18]: Done.

File: apps/hexagon_api/README.md (hunk 2 above)
Comment (kparzysz-quic) [reply apps/hexagon_api/README.md:58]: I updated it with the new variable.

File: apps/hexagon_api/CMakeLists.txt (hunk 1 above)
Comment (mehrdadh) [reply apps/hexagon_api/CMakeLists.txt]: @kparzysz-quic makes sense. Thanks!

**Review by mehrdadh — APPROVED** [review]
@kparzysz-quic I tested with hardware and it LGTM!

**Review by areusch — APPROVED** [review]

**Pull request merged**